### ISSUE TITLE
Rename symbol (F2)

### ISSUE THREAD

**Issue by DonRP** [LuqueDaniel/vscode-language-renpy#169]
https://code.visualstudio.com/docs/editor/refactoring#_rename-symbol

Enabling the renaming of a variable with f2, is a really convenient and safe thing to do (normal replace is very dangerous). but I don't know how complicated it is to implement.

**Comment by JelleInfinity** [comment]
We are planning on updating the semantics and features in next update (after syntax is complete). This update shoul make it a lot more easy to link symbols, and linked symbols would allow rename with F2.

Unfortunately it might take a while for us to start working on this feature as we all don't have much spare time atm.

**Comment by DonRP** [comment]
Yeah, it seems like a long and repetitive job to update the semantics, looking at the file in syntaxes folder. If you want I can help with that.

**Comment by JelleInfinity** [comment]
The plan is to update most of the existing features to use the tokenizer that has already been pushed to the syntax branch.

The next goal is to create a token parser that would build the symbol database and the code graph. I already started some work on this.

Any help would be very much appreciated, I'd say give me a month to finish the above mentioned part and after that it should be easy to use these features to update the rest.

If you'd like to help with something right now, the tokenizer needs a copy of the syntax patterns in typescript. Some of them are already in there, but I didn't update it in a while. I'd love it if you could help me with updating that 😄

**Comment by DonRP** [comment]
OK I can make a few. 

I didn't find a TokenPattern class, but taking a simple one, just turn the json into an object. ex:

```json
"renpy-statements": {
    "patterns": [
        {
        "include": "#label"
        },
        { "include": "#menu" },
        { "include": "#image" },
        { "include": "#audio" },
        { "include": "#transform" },
        { "include": "#scene" },
        { "include": "#camera" },
        { "include": "#show" },
        { "include": "#with" },
        { "include": "#style-old" },
        { "include": "#use-old" },
        { "include": "#screen-old" },
        { "include": "#return-statements" },
        { "include": "#jump" },
        { "include": "#call" }
    ]
},
```
```typescript
const renpyStatements: TokenPattern = {
    patterns: [
        { include: "#label" },
        { include: "#menu" },
        { include: "#image" },
        { include: "#audio" },
        { include: "#transform" },
        { include: "#scene" },
        { include: "#camera" },
        { include: "#show" },
        { include: "#with" },
        { include: "#style-old" },
        { include: "#use-old" },
        { include: "#screen-old" },
        { include: "#return-statements" },
        { include: "#jump" },
        { include: "#call" },
    ]
}
```

**Comment by JelleInfinity** [comment]
Yes, that's about the gist of it. You can use the patterns in the comments of that file to convert them a bit more easily.

Those patterns also apply some changes that are required, eg include pattern in not needed and _name_ is converted to _token_

**Comment by JelleInfinity** [comment]
TokenPattern is a type. You can find it [here](https://github.com/LuqueDaniel/vscode-language-renpy/blob/new-syntax/src/types/token-patterns.d.ts)

**Comment by DonRP** [comment]
OK I am editing it in this fork: https://github.com/DRincs-Productions/vscode-language-renpy

**Comment by DonRP** [comment]
https://github.com/DRincs-Productions/vscode-language-renpy/blob/new-syntax/src/tokenizer/token-patterns.ts

here are some conversions, you notice them because they are the ones with the errors, lol

**Comment by JelleInfinity** [comment]
But thanks for the work you did so far, any bit is already super helpful 🙂
